motoko/basic_bitcoin/src/basic_bitcoin/src/P2trScriptSpend.mo

```diff
@@ -0,0 +1,267 @@
+
+        let script_bytes_sized = Script.toBytes(leaf_script);
+        // remove the size prefix
+        let script_bytes = Array.subArray(script_bytes_sized, 1, script_bytes_sized.size() - 1);
+
+        let _control_block = control_block(tweaked_key_is_even, internal_public_key);
+        // Obtain the scriptPubKey of the source address which is also the
+        // scriptPubKey of the Tx output being spent.
+        switch (Address.scriptPubKey(#p2tr_key own_address)) {
+            case (#ok scriptPubKey) {
+                assert scriptPubKey.size() == 2;
+
+                // Obtain a witness for each Tx input.
+                for (i in Iter.range(0, transaction.txInputs.size() - 1)) {
+                    let sighash = transaction.createTaprootScriptSpendSignatureHash(
+                        amounts,
+                        scriptPubKey,
+                        Nat32.fromIntWrap(i),
+                        leaf_hash,
+                    );
+
+                    Debug.print("Signing sighash: " # debug_show (sighash));
+
+                    let signature = Blob.toArray(await signer(key_name, derivation_path, Blob.fromArray(sighash)));
+                    transaction.witnesses[i] := [signature, script_bytes, _control_block];
+                };
+            };
+            // Verify that our own address is P2TR key spend address.
+            case (#err msg) Debug.trap("This example supports signing p2tr key spend addresses only: " # msg);
+        };
+
+        transaction.toBytes();
+    };
+
+    /// Sends a transaction to the network that transfers the given amount to the
+    /// given destination, where the source of the funds is the canister itself
+    /// at the given derivation path.
+    public func send(network : Network, derivation_path : [[Nat8]], key_name : Text, dst_address : BitcoinAddress, amount : Satoshi) : async [Nat8] {
+        // Get fee percentiles from previous transactions to estimate our own fee.
+        let fee_percentiles = await BitcoinApi.get_current_fee_percentiles(network);
+
+        let fee_per_vbyte : MillisatoshiPerVByte = if (fee_percentiles.size() == 0) {
+            // There are no fee percentiles. This case can only happen on a regtest
+            // network where there are no non-coinbase transactions. In this case,
+            // we use a default of 1000 millisatoshis/vbyte (i.e. 2 satoshi/byte)
+            2000;
+        } else {
+            // Choose the 50th percentile for sending fees.
+            fee_percentiles[50];
+        };
+
+        // Fetch our public key, P2TR script spend address, and UTXOs.
+        let own_sec1_public_key = Blob.toArray(await SchnorrApi.schnorr_public_key(key_name, Array.map(derivation_path, Blob.fromArray)));
+        let own_bip340_public_key = Array.subArray(own_sec1_public_key, 1, 32);
+        let { tweaked_address = own_tweaked_address; is_even } = public_key_to_p2tr_script_spend_address(network, own_bip340_public_key);
```

Comment:
Hmm, maybe I didn't look at this PR for too long, but I don't understand what you exactly mean. Do you mean returning `is_even` also in `get_address`? In that case it would never be used, right? Or did you mean something else?

Reply:
Sorry for being unclear! What I meant was that lines 208-210 here are almost exactly the same as what the function `get_address` does, so I was thinking that we can replace lines 208-210 with a call to `get_address`. For that to work, however, `get_address` would have to be adapted to return the parity `is_even` (which it currently doesn't do) because the parity is here later used in lines 214 and 223.

Reply:
I see. I think it still makes sense to keep the current code instead of replacing it with `get_address` because in that function we are using the raw public key. Although it would be possible to extract the public key from the address text, that would still be a complication in some other place instead of this one, if we provide a function for extracting the public key. WDYT?

Reply:
Ah, my bad! I missed that we are not only using `own_tweaked_address` but also `own_bip340_public_key` later. Sorry for the noise!